### PR TITLE
LLVM: Fix `list.insert()` when given index is greater than length of the list

### PR DESCRIPTION
**Pull request by faze-geek**
Consider `list.insert(position,element)`, In python if `position `is greater than the length of list then by default the `element `is inserted at `position = len(list)`. 
On master - 
```
def f():
    l:list[str] = ["a","b","c","d"]
    l.insert(6,"e")
    print(l)
f()

(lp) faze-geek@DESKTOP-497JDCU:~/lpython/lpython$ python try.py
['a', 'b', 'c', 'd', 'e']
(lp) faze-geek@DESKTOP-497JDCU:~/lpython/lpython$ ./src/bin/lpython try.py
['a', 'b', 'c', 'd', '(null)']    #lpython adds a garbage value

def f():
    l:list[i32] = [1,2,3]
    l.insert(8,4)
    print(l)
f()

(lp) faze-geek@DESKTOP-497JDCU:~/lpython/lpython$ python try.py
[1, 2, 3, 4]
(lp) faze-geek@DESKTOP-497JDCU:~/lpython/lpython$ ./src/bin/lpython try.py
[1, 2, 3, 0]    #lpython adds a garbage value

```

On branch - 
```
def f():
    l:list[str] = ["a","b","c","d"]
    l.insert(6,"e")
    print(l)
f()

(lp) C:\Users\kunni\lpython>python try.py
['a', 'b', 'c', 'd', 'e']
(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
['a', 'b', 'c', 'd', 'e']
```

### CODE REVIEW

File: src/libasr/codegen/llvm_utils.cpp [hunk 1]
```diff
@@ -2365,8 +2365,18 @@ namespace LCompilers {
...
+            builder->CreateICmpSGT(pos, current_end_point),
+            current_end_point,
+            pos
+        );
+        adjusted_pos = builder->CreateSelect(
+            builder->CreateICmpSGE(adjusted_pos, llvm::ConstantInt::get(context, llvm::APInt(32, 0))),
+            adjusted_pos,
+            llvm::ConstantInt::get(context, llvm::APInt(32, 0))
+        );
         llvm::AllocaInst *tmp_ptr = builder->CreateAlloca(el_type, nullptr);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/llvm_utils.cpp:2378]: The TODO at line 2365 is for AllocaInst, I think. Why don't you move this above?

Comment (faze-geek) [reply src/libasr/codegen/llvm_utils.cpp:2378]: Okay sure. You are right.

File: src/libasr/codegen/llvm_utils.cpp [hunk 2]
```diff
@@ -2344,7 +2344,17 @@ namespace LCompilers {
...
+            builder->CreateICmpSGT(pos, current_end_point),
+            current_end_point,
+            pos
+        );
+        adjusted_pos = builder->CreateSelect(
+            builder->CreateICmpSGE(adjusted_pos, llvm::ConstantInt::get(context, llvm::APInt(32, 0))),
+            adjusted_pos,
+            llvm::ConstantInt::get(context, llvm::APInt(32, 0))
+        );
+
```
Comment (czgdp1807) [inline src/libasr/codegen/llvm_utils.cpp:2357]: I guess so for this, we can use, [CreateMinimum](https://llvm.org/doxygen/classllvm_1_1IRBuilderBase.html#ada46861dd8cf60fa7e1ac61bbd385b49). Its just like, `min(len(list), position)`. If `position` will be greater than `len(list)` then automatically `position` will not be used and otherwise `position` will be used.

Comment (faze-geek) [reply src/libasr/codegen/llvm_utils.cpp:2357]: Sorry I have not got very familiar with `llvm` , I tried - 
`llvm::Value* adjusted_pos = builder->CreateMinimum(pos, current_end_point);
`
But it leads to an error while running above test cases - 
```
attributes #0 = { argmemonly nounwind willreturn }
attributes #1 = { nounwind readnone speculatable willreturn }
code generation error: asr_to_llvm: module failed verification. Error:
Intrinsic has incorrect return type!
i32 (i32, i32)* @llvm.minimum.i32
```
Could you help me out here !

Comment (czgdp1807) [reply src/libasr/codegen/llvm_utils.cpp:2357]: I see. Why are we having, `adjusted_pos = builder->CreateSelect(
             builder->CreateICmpSGE(adjusted_pos, llvm::ConstantInt::get(context, llvm::APInt(32, 0))),
             adjusted_pos,
             llvm::ConstantInt::get(context, llvm::APInt(32, 0))
         );`?  For handling negative indices?

Comment (faze-geek) [reply src/libasr/codegen/llvm_utils.cpp:2357]: @czgdp1807  Sorry I could not respond to this earlier, was busy with exams at university. 

Yes, that's correct. The code you mentioned is responsible for handling negative indices. Basically, it ensures new item is inserted at the beginning of the list if a negative index is provided.

Comment (faze-geek) [reply src/libasr/codegen/llvm_utils.cpp:2357]: I remember a comment regarding not allowing negative indexing in the list ( allowed as of now ). You can check it [here](https://github.com/lcompilers/lpython/issues/1533#issuecomment-1445144248)
 
It thats the case I'll just remote the snippet checking for negative indices. Rest the pr looks good to me.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]
Please add a test for it or point me to any existing tests which verify that your change works correctly.

**Comment by faze-geek** [comment]
> Please add a test for it or point me to any existing tests which verify that your change works correctly.

I have added a test case now. Basically, it handles cases when first argument is greater than list length.

File: integration_tests/test_list_03.py [hunk 3]
```diff
@@ -24,6 +24,11 @@ def test_list_insert_02(x: list[i32], n: i32) -> list[i32]:
 
     return x
 
+def test_list_insert_03():
+    l1:list[str] = ["a","b","c","d"]
+    l1.insert(6,"e")
+    assert l1 == ['a', 'b', 'c', 'd', 'e']
+
```
Comment (czgdp1807) [inline integration_tests/test_list_03.py:31]:
```suggestion
    l1:list[str] = ["a", "b", "c", "d"]
    l1.insert(6,"e")
    assert l1 == ['a',  'b',  'c',  'd',  'e']
    l1.insert(18, "f")
    assert l1 == ['a', 'b', 'c', 'd', 'e', 'f']
```

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: integration_tests/test_list_03.py (hunk 3 above)
Comment (faze-geek) [reply integration_tests/test_list_03.py:31]: Okay I'll add this quick. 
Also the logs fail, does asserting like this work or should I try asserting each element in a for loop ?

Comment (czgdp1807) [reply integration_tests/test_list_03.py:31]: Asserting each element should be better option for making sure our list algorithms are working as expected. But as a feature direct comparison of lists should also work (but that's a separate issue).

Comment (faze-geek) [reply integration_tests/test_list_03.py:31]: As of now, all the integration tests do element wise comparison (possibly because direct assertion is an issue). I'll use a for loop now.

Comment (faze-geek) [reply integration_tests/test_list_03.py]: I have updated the test accordingly.

**Comment by faze-geek** [comment]
@czgdp1807 
I investigated the failing logs now. Firstly the running `integration_tests/test_list_03.py` (where tests have been added) does not give a segmentation error in windows MSVC unlike my WSL setup. This is a seperate issue.
Secondly the `c backend` in WSL raises the Segmentation Fault when I test indexes greater than the length of the list and as I use an index within the length the error does not occur.
But then this case is a valid edge case as I have shown above, comparing with Cpython -
```
def test_list_insert_03():
    l1:list[str] = ["a", "b", "c", "d"]
    l1.insert(5,"e")
    s:str = "a"
    i:i32 = 0
    # for i in range(len(l1)):
    #     assert l1[i] == chr(ord(s)+i)
    print(l1)

test_list_insert_03()

(lp) C:\Users\kunni\lpython>python try.py
['a', 'b', 'c', 'd', 'e']

(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
['a', 'b', 'c', 'd', 'e'] 
```

File: src/libasr/codegen/llvm_utils.cpp (hunk 2 above)
Comment (czgdp1807) [inline src/libasr/codegen/llvm_utils.cpp:2357]: Seems like an issue with edge case. Note that `current_end_point` equals `len(list)` all the times.

https://github.com/lcompilers/lpython/blob/57ebbc81e23854d3af50c2d5042f2f1b184941b5/src/libasr/codegen/llvm_utils.cpp#L2252-L2254
